File: sdk/python/symbolchain/symbol/MessageEncoder.py
import warnings
from binascii import Error as BinasciiError
from binascii import hexlify, unhexlify

import cryptography

from symbolchain.CryptoTypes import PrivateKey, PublicKey
from symbolchain.impl.CipherHelpers import decode_aes_gcm, encode_aes_gcm
from symbolchain.symbol.KeyPair import KeyPair
from symbolchain.symbol.SharedKey import SharedKey

DELEGATION_MARKER = unhexlify('FE2A8061577301E2')


class MessageEncoder:
	"""Encrypts and encodes messages between two parties."""

	def __init__(self, key_pair: KeyPair):
		"""Creates message encoder around key pair."""

		self.key_pair = key_pair

	def try_decode(self, recipient_public_key, encoded_message):
		"""Tries to decode encoded message, returns tuple:
		* True, message - if message has been decoded and decrypted
		* False, encoded_message - otherwise
		"""

		if 1 == encoded_message[0]:
			try:
				message = decode_aes_gcm(SharedKey, self.key_pair, recipient_public_key, encoded_message[1:])
				return True, message
			except cryptography.exceptions.InvalidTag:
				pass
		elif 0xFE == encoded_message[0] and DELEGATION_MARKER == encoded_message[:8]:
			try:
				ephemeral_public_key = PublicKey(encoded_message[len(DELEGATION_MARKER):len(DELEGATION_MARKER) + PublicKey.SIZE])
				message = decode_aes_gcm(SharedKey, self.key_pair, ephemeral_public_key, encoded_message[len(DELEGATION_MARKER) + PublicKey.SIZE:])
				return True, message
			except cryptography.exceptions.InvalidTag:
				pass
			except ValueError as exception:
				exceptions = ['point is not in main subgroup']
				if not any(map(lambda message: message in str(exception), exceptions)):
					raise

		return False, encoded_message

	def encode(self, recipient_public_key: PublicKey, message: bytes):
		"""Encodes message to recipient using recommended format."""

		tag, initialization_vector, cipher_text = encode_aes_gcm(SharedKey, self.key_pair, recipient_public_key, message)
		return b'\1' + tag + initialization_vector + cipher_text

	@staticmethod
	def encode_persistent_harvesting_delegation(node_public_key, remote_key_pair, vrf_root_key_pair):  # pylint: disable=invalid-name
		"""Encodes persistent harvesting delegation to node."""

		ephemeral_key_pair = KeyPair(PrivateKey.random())

		message = remote_key_pair.private_key.bytes + vrf_root_key_pair.private_key.bytes
		tag, initialization_vector, cipher_text = encode_aes_gcm(SharedKey, ephemeral_key_pair, node_public_key, message)

		return DELEGATION_MARKER + ephemeral_key_pair.public_key.bytes + tag + initialization_vector + cipher_text

	def try_decode_deprecated(self, recipient_public_key, encoded_message):
		"""Tries to decode encoded wallet message, returns tuple:
		* True, message - if message has been decoded and decrypted
		* False, encoded_message - otherwise
		"""

		warnings.warn(
			'This function is only provided for compatability with the original Symbol wallets.\n'
			'Please use `try_decode` in any new code.',
			category=DeprecationWarning,
			stacklevel=2)

		if 1 == encoded_message[0]:
			try:
				# wallet additionally hex encodes
				return self.try_decode(recipient_public_key, bytes([1]) + unhexlify(encoded_message[1:].decode('utf8')))
			except (UnicodeDecodeError, BinasciiError):
				pass

		return self.try_decode(recipient_public_key, encoded_message)

	def encode_deprecated(self, recipient_public_key: PublicKey, message: bytes):
		"""Encodes message to recipient using (deprecated) wallet format."""

		warnings.warn(
			'This function is only provided for compatability with the original Symbol wallets.\n'
			'Please use `encode` in any new code.',
			category=DeprecationWarning,
			stacklevel=2)

		# wallet additionally hex encodes
		return bytes([1]) + hexlify(self.encode(recipient_public_key, message)[1:])

File: sdk/javascript/src/symbol/MessageEncoder.js
import { KeyPair } from './KeyPair.js';
import { deriveSharedKey } from './SharedKey.js';
import { PrivateKey, PublicKey } from '../CryptoTypes.js';
import { concatArrays, decodeAesGcm, encodeAesGcm } from '../impl/CipherHelpers.js';
import { deepCompare } from '../utils/arrayHelpers.js';
import { hexToUint8, isHexString, uint8ToHex } from '../utils/converter.js';

const DELEGATION_MARKER = Uint8Array.from(Buffer.from('FE2A8061577301E2', 'hex'));

const filterExceptions = (statement, exceptions) => {
	try {
		const message = statement();
		return [true, message];
	} catch (exception) {
		if (!exceptions.some(exceptionMessage => exception.message.includes(exceptionMessage)))
			throw exception;
	}

	return [false, undefined];
};

/**
 * Encrypts and encodes messages between two parties.
 */
export default class MessageEncoder {
	/**
	 * Creates message encoder around key pair.
	 * @param {KeyPair} keyPair Key pair.
	 */
	constructor(keyPair) {
		/**
		 * @private
		 */
		this._keyPair = keyPair;
	}

	/**
	 * Public key used for message encoding.
	 * @returns {PublicKey} Public key used for message encoding.
	 */
	get publicKey() {
		return this._keyPair.publicKey;
	}

	/**
	 * Tries to decode encoded message.
	 * @param {PublicKey} recipientPublicKey Recipient's public key.
	 * @param {Uint8Array} encodedMessage Encoded message.
	 * @returns {TryDecodeResult} Tuple containing decoded status and message.
	 */
	tryDecode(recipientPublicKey, encodedMessage) {
		if (1 === encodedMessage[0]) {
			const [result, message] = filterExceptions(
				() => decodeAesGcm(deriveSharedKey, this._keyPair, recipientPublicKey, encodedMessage.subarray(1)),
				['Unsupported state or unable to authenticate data']
			);
			if (result)
				return { isDecoded: true, message };
		}

		if (0xFE === encodedMessage[0] && 0 === deepCompare(DELEGATION_MARKER, encodedMessage.slice(0, 8))) {
			const ephemeralPublicKeyStart = DELEGATION_MARKER.length;
			const ephemeralPublicKeyEnd = ephemeralPublicKeyStart + PublicKey.SIZE;
			const ephemeralPublicKey = new PublicKey(encodedMessage.subarray(ephemeralPublicKeyStart, ephemeralPublicKeyEnd));

			const [result, message] = filterExceptions(
				() => decodeAesGcm(deriveSharedKey, this._keyPair, ephemeralPublicKey, encodedMessage.subarray(ephemeralPublicKeyEnd)),
				[
					'Unsupported state or unable to authenticate data',
					'invalid point'
				]
			);
			if (result)
				return { isDecoded: true, message };
		}

		return { isDecoded: false, message: encodedMessage };
	}

	/**
	 * Encodes message to recipient using recommended format.
	 * @param {PublicKey} recipientPublicKey Recipient public key.
	 * @param {Uint8Array} message Message to encode.
	 * @returns {Uint8Array} Encrypted and encoded message.
	 */
	encode(recipientPublicKey, message) {
		const { tag, initializationVector, cipherText } = encodeAesGcm(deriveSharedKey, this._keyPair, recipientPublicKey, message);

		return concatArrays(new Uint8Array([1]), tag, initializationVector, cipherText);
	}

	/**
	 * Encodes persistent harvesting delegation to node.
	 * @param {PublicKey} nodePublicKey Node public key.
	 * @param {KeyPair} remoteKeyPair Remote key pair.
	 * @param {KeyPair} vrfKeyPair Vrf key pair.
	 * @returns {Uint8Array} Encrypted and encoded harvesting delegation request.
	 */
	// eslint-disable-next-line class-methods-use-this
	encodePersistentHarvestingDelegation(nodePublicKey, remoteKeyPair, vrfKeyPair) {
		const ephemeralKeyPair = new KeyPair(PrivateKey.random());
		const message = concatArrays(remoteKeyPair.privateKey.bytes, vrfKeyPair.privateKey.bytes);
		const { tag, initializationVector, cipherText } = encodeAesGcm(deriveSharedKey, ephemeralKeyPair, nodePublicKey, message);

		return concatArrays(DELEGATION_MARKER, ephemeralKeyPair.publicKey.bytes, tag, initializationVector, cipherText);
	}

	/**
	 * Tries to decode encoded message.
	 * @deprecated This function is only provided for compatability with the original Symbol wallets.
	 *             Please use `tryDecode` in any new code.
	 * @param {PublicKey} recipientPublicKey Recipient's public key.
	 * @param {Uint8Array} encodedMessage Encoded message
	 * @returns {TryDecodeResult} Tuple containing decoded status and message.
	 */
	tryDecodeDeprecated(recipientPublicKey, encodedMessage) {
		const encodedHexString = new TextDecoder().decode(encodedMessage.subarray(1));
		if (1 === encodedMessage[0] && isHexString(encodedHexString)) {
			// wallet additionally hex encodes
			return this.tryDecode(recipientPublicKey, new Uint8Array([1, ...hexToUint8(encodedHexString)]));
		}

		return this.tryDecode(recipientPublicKey, encodedMessage);
	}

	/**
	 * Encodes message to recipient using (deprecated) wallet format.
	 * @deprecated This function is only provided for compatability with the original Symbol wallets.
	 *             Please use `encode` in any new code.
	 * @param {PublicKey} recipientPublicKey Recipient public key.
	 * @param {Uint8Array} message Message to encode.
	 * @returns {Uint8Array} Encrypted and encoded message.
	 */
	encodeDeprecated(recipientPublicKey, message) {
		// wallet additionally hex encodes
		const encodedHexString = uint8ToHex(this.encode(recipientPublicKey, message).subarray(1));
		const encodedHexStringBytes = new TextEncoder().encode(encodedHexString);
		return new Uint8Array([1, ...encodedHexStringBytes]);
	}
}

// region type declarations

/**
 * Result of a try decode operation.
 * @class
 * @typedef {object} TryDecodeResult
 * @property {boolean} isDecoded \c true if message has been decoded and decrypted; \c false otherwise.
 * @property {Uint8Array} message Decoded message when `isDecoded` is \c true; encoded message otherwise.
 */

// endregion
